freight/providers/pipeline.py

```diff
@@ -0,0 +1,359 @@
+                namespace=deployment.metadata.namespace,
+                body=deployment,
+            )
+
+            watchers.append(
+                (
+                    partial(
+                        rollout_status_deployment,
+                        api,
+                        resp.metadata.name,
+                        resp.metadata.namespace,
+                        resp.metadata.generation,
+                    ),
+                    {},  # empty state dict for this rollout
+                )
+            )
+
+    return watchers
+
+
+def run_step_shell(step: dict, context: PipelineContext):
+    command = format_task(step["command"], context.task)
+    context.workspace.run(command, env=step.get("env"))
+
+
+def format_task(data, task: TaskContext):
+    if isinstance(data, str):
+        return data.format(**asdict(task))
+    if isinstance(data, list):
+        return [format_task(d, task) for d in data]
+    if isinstance(data, dict):
+        return {k: format_task(v, task) for k, v in data.items()}
+    return data
+
+
+def make_job_spec(step: dict, task: TaskContext) -> dict:
+    return {
+        "kind": "Job",
+        "metadata": {
+            "name": f"{step['name']}-{int(time())}",
+            "namespace": step.get("namespace", "default"),
+        },
+        "spec": {
+            "template": {
+                "spec": {
+                    "containers": [
+                        {
+                            "name": "job",
+                            "image": format_task(step["image"], task),
+                            "args": format_task(step.get("args", []), task),
+                            "env": format_task(step.get("env", []), task),
+                            "volumeMounts": step.get("volumeMounts", []),
+                            "resources": step.get("resources", {}),
+                        }
+                    ],
+                    "volumes": step.get("volumes", []),
+                    "restartPolicy": "Never",
+                }
+            },
+            "backoffLimit": 0,
+        },
+    }
+
+
+def run_step_job(step: dict, context: PipelineContext):
+    context.workspace.log.info(f"Running Job: {repr(step)}")
+
+    job_spec = make_job_spec(step, context.task)
+    name = job_spec["metadata"]["name"]
+    namespace = job_spec["metadata"]["namespace"]
+
+    context.workspace.log.info(f"Creating Job {namespace}/{name}")
+    client.BatchV1Api(context.kube.client).create_namespaced_job(
+        namespace=namespace, body=job_spec
+    )
+    pod_name = None
+    try:
+        pods = (
+            client.CoreV1Api(context.kube.client)
+            .list_namespaced_pod(namespace=namespace, label_selector=f"job-name={name}")
+            .items
+        )
+        assert len(pods) == 1, len(pods)
+        pod_name = pods[0].metadata.name
+        context.workspace.log.info(f"Waiting for Pod {pod_name}")
+        read_logs = False
```

Comment:
minor; this loop was pretty confusing to read my first time around, perhaps we have a comment here explaining `read_logs` and why we loop (e.g. loop around to get the final pod status and check failure)?

Reply:
I was in the process of adding comments to this function since it's the most complex, so yeah. Lemme finish that up.